mlir/lib/Bytecode/Writer/BytecodeWriter.cpp

```diff
@@ -1256,6 +1288,21 @@ void BytecodeWriter::writePropertiesSection(EncodingEmitter &emitter) {
 
 LogicalResult mlir::writeBytecodeToFile(Operation *op, raw_ostream &os,
                                         const BytecodeWriterConfig &config) {
+  // Before creating the bytecode writer, give an opportunity to each of the
+  // dialects that registered a target version emission to perform IR
+  // downgrades. Note that it is responsibility of client dialect to handle the
+  // potentially destructive changes that a downgrade function could do to the
+  // IR.
+  for (auto &item : config.getDialectVersionMap()) {
+    Dialect *currentDialect = op->getContext()->getOrLoadDialect(item.getKey());
+    assert(currentDialect &&
+           "version requested on a dialect that could not be loaded");
+    auto *iface = llvm::dyn_cast<BytecodeDialectInterface>(currentDialect);
+    assert(iface && "version requested on a dialect that did not implement a "
+                    "`BytecodeDialectInterface`");
+    if (failed(iface->downgradeToVersion(op, *item.getValue())))
+      return failure();
```

Comment:
I don't think `writeBytecodeToFile` should be allowed to modify the input IR in any way actually.

Reply:
Would you suggest cloning instead? The issue I've ran into is that cloning before doing IR downgrades will not maintain use-lists chains, which probably suggests an issue with op->clone() that we should fix.

Reply:
That seems to me like a legalization pass pre-serialization instead?

Reply:
It can be done in a separate, pre-serialization pass.

What made me lean towards baking this within the writeBytecodeToFile call is that pre-serialization would require you to handle the target emission version in two places: first, your downgrade, pre serialization pass. Second, your serialization call (needed to downgrade type/attr/property encodings). I find this split not really necessary, but maybe I am missing some other aspects you have in mind? The current implementation gives you the ability to set the target emission version into the writer config only once, and that's it - it all happens under the hood, somewhat mirroring the read-and-upgrade approach we implemented few months ago. I can change this though, and make the write call somewhat leaner if that's confusing.

Reply:
I feel strongly that it is unusual and misleading that a "write" or "print" API would be modifying the module.
That is if our classes were const-correct, the "writeByteCodeToFile" (just like "print") API should take a `const Operation *`.

Reply:
Thanks for the feedback. I removed the functionality associated to downgrading constructs, which is now left to the client entirely, and added an example of downgrade as part of the tests. I also modified title and description accordingly.